accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -579,6 +617,153 @@ all occurrences of `S*` in `T` shall be replaced with `S`.  As a result, legacy
+are not all syntactically equal.  For the purposes of runtime subtyping checks,
+`C` is considered to implement the canonical `LEGACY_ERASURE` of the
+super-interfaces in question.
+
+A member which is defined in a class in a legacy library (whether concrete or
+abstract), is given a signature in which every type is a legacy type.  It is an
+error if the signature of a member is not a correct override of all members of
+the same name in super-interfaces of the class, using the legacy subtyping
+rules.
+
+#### Classes defined in legacy libraries as seen from opted-in libraries
+
+The `LEGACY_TOP_MERGE` of two types `T` and `S` is the unique type `R` defined
+as:
+ - `LEGACY_TOP_MERGE(T?, S?) = LEGACY_TOP_MERGE(T, S)?`
+ - `LEGACY_TOP_MERGE(T?, S*) = LEGACY_TOP_MERGE(T, S)*`
+ - `LEGACY_TOP_MERGE(T*, S?) = LEGACY_TOP_MERGE(T, S)*`
+ - `LEGACY_TOP_MERGE(T*, S*) = LEGACY_TOP_MERGE(T, S)*`
+ - `LEGACY_TOP_MERGE(T*, S)  = LEGACY_TOP_MERGE(T, S)*`
+ - `LEGACY_TOP_MERGE(T, S*)  = LEGACY_TOP_MERGE(T, S)*`
+ - `LEGACY_TOP_MERGE(T?, S)  = LEGACY_TOP_MERGE(T, S)*`
+ - `LEGACY_TOP_MERGE(T, S?)  = LEGACY_TOP_MERGE(T, S)*`
+ - `LEGACY_TOP_MERGE(Never, Null)  = Null`
+ - `LEGACY_TOP_MERGE(Null, Never)  = Null`
+ - `LEGACY_TOP_MERGE(void, void)  = void`
+ - `LEGACY_TOP_MERGE(Object, Object)  = Object`
+ - `LEGACY_TOP_MERGE(dynamic, dynamic)  = Object`
+ - `LEGACY_TOP_MERGE(Object, void)  = void`
+   - And the reverse
+ - `LEGACY_TOP_MERGE(dynamic, void)  = void`
+   - And the reverse
+ - `LEGACY_TOP_MERGE(Object, dynamic)  = Object`
+   - And the reverse
+ - And for all other types, recursively applying the transformation over the
+   structure of the type
+   - e.g. `LEGACY_TOP_MERGE(C<T>, C<S>)  = C<LEGACY_TOP_MERGE(T, S)>`
+
+In other words, `LEGACY_TOP_MERGE` takes two types which are structurally equal
+except for the placement of `?` and `*`, the use of `Null` vs `Never`, and the
+particular choice of top types, and finds a single canonical type to represent
+them. The `LEGACY_TOP_MERGE` of two types is not always defined.
+
+If a class which is defined in a legacy library inherits a member with the same
+name from multiple super-interfaces, then error checking is done as usual using
+the legacy typing rules which ignore nullability.  This means that it is valid
+for a legacy class to inherit the same member signature with contradictory
+nullability information. For the purposes of member lookup within the legacy
+library itself, nullability information is ignored, and so it is valid to simply
+erase the nullability information within the legacy library. When referenced
+from an opted-in library, if there are multiple signatures `T0, ... Tn` for a
+member `m`, and there is a single `Ti` such that `LEGACY_SUBTYPE(Ti, Tk)` for
+each `k` in `0, ..., n`, then the signature of `m` is taken as `Ti`.  Otherwise,
+the signature of `m` for the purposes of member lookup is the `LEGACY_TOP_MERGE`
+of `S0, ..., Sn`, where `Si` is **NORM(`Ti`)**.
```

Comment:
`LEGACY_TOP_MERGE` is defined for two types, does applying it to more then two types mean applying to the first two types, then to the result and the third type, etc?

Reply:
Yes, it should be commutative and associative.